test/e2e/network/traffic_policy.go

```diff
@@ -0,0 +1,476 @@
+
+	var finalSlices []discoveryv1.EndpointSlice
+	err = wait.Poll(5*time.Second, svcReadyTimeout, func() (bool, error) {
+		slices, listErr := c.DiscoveryV1().EndpointSlices(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", discoveryv1.LabelServiceName, svc.Name)})
+		if listErr != nil {
+			return false, listErr
+		}
+
+		numEndpoints := 0
+		for _, slice := range slices.Items {
+			numEndpoints += len(slice.Endpoints)
+		}
+		if len(schedulableNodes) > numEndpoints {
+			framework.Logf("Expected %d endpoints, got %d", len(schedulableNodes), numEndpoints)
+			return false, nil
+		}
+
+		finalSlices = slices.Items
+		return true, nil
+	})
+	framework.ExpectNoError(err, "timed out waiting for EndpointSlices to be ready")
+
+	ginkgo.By("having correct hints set for each endpoint")
+	for _, slice := range finalSlices {
+		for _, ep := range slice.Endpoints {
+			if ep.Zone == nil {
+				framework.Failf("Expected endpoint in %s to have zone: %v", slice.Name, ep)
+			}
+			if topologyHints {
+				if ep.Hints == nil || len(ep.Hints.ForZones) == 0 {
+					framework.Failf("Expected endpoint in %s to have hints: %v", slice.Name, ep)
+				}
+				if len(ep.Hints.ForZones) > 1 {
+					framework.Failf("Expected endpoint in %s to have exactly 1 zone hint, got %d: %v", slice.Name, len(ep.Hints.ForZones), ep)
+				}
+				if *ep.Zone != ep.Hints.ForZones[0].Name {
+					framework.Failf("Expected endpoint in %s to have same zone hint, got %s: %v", slice.Name, *ep.Zone, ep)
+				}
+			} else {
+				if ep.Hints != nil && len(ep.Hints.ForZones) > 0 {
+					framework.Failf("Unexpected hints on endpoint in %s: %v", slice.Name, ep)
+				}
+			}
+
+		}
+	}
+
+	nodesByZone := make(map[string]string)
+	zonesWithNodes := make(map[string][]*v1.Node)
+	for _, node := range schedulableNodes {
+		if zone, ok := node.Labels[v1.LabelTopologyZone]; ok {
+			nodesByZone[node.Name] = zone
+			zonesWithNodes[zone] = append(zonesWithNodes[zone], node)
+		}
+	}
+
+	podList, err := c.CoreV1().Pods(f.Namespace.Name).List(context.TODO(), metav1.ListOptions{})
+	framework.ExpectNoError(err)
+	for _, pod := range podList.Items {
+		if zone, ok := nodesByZone[pod.Spec.NodeName]; ok {
+			tr.zonesByPod[pod.Name] = zone
+			tr.nodesByPod[pod.Name] = pod.Spec.NodeName
+		}
+	}
+
+	// use the largest zone to determine which node's port to use for nodeport endpoint requests
+	largestSeen := 0
+	for zone, nodes := range zonesWithNodes {
+		if len(nodes) > largestSeen {
+			largestSeen = len(nodes)
+			tr.endpointZone = zone
+		}
+	}
+	framework.ExpectNotEqual("", tr.endpointZone, "unable to find a zone with more than one node")
+
+	// always send to a node in the largest zone so balancing will result in more than 1 host name being seen
+	tr.endpointNode = zonesWithNodes[tr.endpointZone][0]
+	// same zone as endpointNode
+	peerNode := zonesWithNodes[tr.endpointZone][1]
```

Comment:
Are we guaranteed to have more than one node in at least one zone? If not, it seems like this would break?

Reply:
I think you're right, good call. It probably makes sense to check largestSeen > 1 at line 226.

Reply:
I added a check and skip in the latest update. Verified with kind locally. See line 228